### PR TITLE
mlx5: Expose DV APIs for direct rule managing

### PR DESCRIPTION
**Pull request by yishaih**
This series exposes mlx5 DV APIs for direct rule managing to increase insertion rate as of bypassing the kernel and the firmware.

A detailed man page was added to describe the expected usage.

The 5 first patches from this series are from other PR (i.e. Add DM support) which depends on some kernel code to be accepted, the other patches have no dependency on any kernel code.

### CODE REVIEW

File: providers/mlx5/dr_send.c [hunk 1]
```diff
@@ -0,0 +1,1039 @@
...
+	qp_create_attr.page_id		= attr->uar->page_id;
+	qp_create_attr.pdn		= attr->pdn;
+	qp_create_attr.cqn		= attr->cqn;
+	qp_create_attr.pm_state		= MLX5_QPC_PM_STATE_MIGRATED;
+	qp_create_attr.service_type	= MLX5_QPC_ST_RC;
+	qp_create_attr.buff_umem_id	= dr_qp->buf_umem->umem_id;
+	qp_create_attr.db_umem_id	= dr_qp->db_umem->umem_id;
+	qp_create_attr.sq_wqe_cnt	= dr_qp->sq.wqe_cnt;
+	qp_create_attr.rq_wqe_cnt	= dr_qp->rq.wqe_cnt;
+	qp_create_attr.rq_wqe_shift	= dr_qp->rq.wqe_shift;
```
Comment (jgunthorpe) [inline providers/mlx5/dr_send.c]: I would be happy if much this code were run through clang-format to get rid of stuff like this..

File: providers/mlx5/dr_ste.c [hunk 2]
```diff
@@ -0,0 +1,2409 @@
...
+#define STE_IPV4		0x1
+#define STE_IPV6		0x2
+#define STE_TCP			0x1
+#define STE_UDP			0x2
+#define STE_SPI			0x3
+#define IP_VERSION_IPV4		0x4
+#define IP_VERSION_IPV6		0x6
+#define IP_PROTOCOL_UDP		0x11
+#define IP_PROTOCOL_TCP		0x06
+#define IP_PROTOCOL_IPSEC	0x33
```
Comment (jgunthorpe) [inline providers/mlx5/dr_ste.c]: This horizontal white space doesn't even line up...

File: providers/mlx5/dr_ste.c [hunk 3]
```diff
@@ -0,0 +1,2409 @@
...
+
+/* Read from layout struct */
+#define DR_STE_GET(typ, p, fld) DEVX_GET(ste_##typ, p, fld)
+
+/* Write to layout a value */
+#define DR_STE_SET(typ, p, fld, v) DEVX_SET(ste_##typ, p, fld, v)
+
+/* Set to STE a specific value using DR_STE_SET */
+#define DR_STE_SET_VAL(lookup_type, tag, t_fname, spec, s_fname, value) do { \
+	if (spec->s_fname) { \
```
Comment (jgunthorpe) [inline providers/mlx5/dr_ste.c]: brackets in macros again

File: providers/mlx5/dr_ste.c [hunk 4]
```diff
@@ -0,0 +1,2409 @@
...
+#define DR_STE_SET_TCP_FLAGS(lookup_type, tag, spec) do { \
+	DR_STE_SET(lookup_type, tag, tcp_ns, !!(spec->tcp_flags & (1 << 8))); \
+	DR_STE_SET(lookup_type, tag, tcp_cwr, !!(spec->tcp_flags & (1 << 7))); \
+	DR_STE_SET(lookup_type, tag, tcp_ece, !!(spec->tcp_flags & (1 << 6))); \
+	DR_STE_SET(lookup_type, tag, tcp_urg, !!(spec->tcp_flags & (1 << 5))); \
+	DR_STE_SET(lookup_type, tag, tcp_ack, !!(spec->tcp_flags & (1 << 4))); \
+	DR_STE_SET(lookup_type, tag, tcp_psh, !!(spec->tcp_flags & (1 << 3))); \
+	DR_STE_SET(lookup_type, tag, tcp_rst, !!(spec->tcp_flags & (1 << 2))); \
+	DR_STE_SET(lookup_type, tag, tcp_syn, !!(spec->tcp_flags & (1 << 1))); \
+	DR_STE_SET(lookup_type, tag, tcp_fin, !!(spec->tcp_flags & (1 << 0))); \
```
Comment (jgunthorpe) [inline providers/mlx5/dr_ste.c]: The !! are so ugly. I think you want a DR_STE_SET_BOOL

File: providers/mlx5/dr_ste.c [hunk 5]
```diff
@@ -0,0 +1,2409 @@
...
+		       in_out##_first_mpls_exp); \
+	DR_STE_SET_TAG(lookup_type, tag, mpls0_ttl, mask, \
+		       in_out##_first_mpls_ttl); \
+} while (0)
+
+#define DR_STE_IS_OUTER_MPLS_OVER_GRE_SET(_misc) (\
+	!!_misc->outer_first_mpls_over_gre_label || \
+	!!_misc->outer_first_mpls_over_gre_exp || \
+	!!_misc->outer_first_mpls_over_gre_s_bos || \
+	!!_misc->outer_first_mpls_over_gre_ttl)
```
Comment (jgunthorpe) [inline providers/mlx5/dr_ste.c]: More redundant !! and missing ()

File: providers/mlx5/dr_ste.c [hunk 6]
```diff
@@ -0,0 +1,2409 @@
...
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 3, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 4, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 5, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 6, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 7, i);
+	}
+}
+
+/* Compute CRC32 (Slicing-by-8 algorithm) */
+static uint32_t dr_crc32_slice8_calc(const void *input_data, size_t length)
```
Comment (jgunthorpe) [inline providers/mlx5/dr_ste.c]: Is this a generic crc32? Would be happier to just use standard crc stuff from ccan:

https://github.com/rustyrussell/ccan/tree/master/ccan/crc

File: providers/mlx5/dr_ste.c [hunk 7]
```diff
@@ -0,0 +1,2409 @@
...
+		dr_ste_always_miss_addr(&ste, connect_info->miss_icm_addr);
+}
+
+int dr_ste_htbl_init_and_postsend(struct mlx5dv_dr_domain *dmn,
+				  struct dr_domain_rx_tx *nic_dmn,
+				  struct dr_ste_htbl *htbl,
+				  struct dr_htbl_connect_info *connect_info,
+				  bool update_hw_ste)
+{
+	uint8_t formated_ste[DR_STE_SIZE] = {0};
```
Comment (jgunthorpe) [inline providers/mlx5/dr_ste.c]: avoid {0}

File: providers/mlx5/dr_ste.c [hunk 8]
```diff
@@ -0,0 +1,2409 @@
...
+				      enum dr_icm_chunk_size chunk_size,
+				      uint8_t lu_type, uint16_t byte_mask)
+{
+	struct dr_icm_chunk *chunk;
+	struct dr_ste_htbl *htbl;
+	int i;
+
+	htbl = calloc(1, sizeof(struct dr_ste_htbl));
+	if (!htbl) {
+		errno = ENOMEM;
```
Comment (jgunthorpe) [inline providers/mlx5/dr_ste.c]: errno is redundant

File: providers/mlx5/dr_ste.c [hunk 9]
```diff
@@ -0,0 +1,2409 @@
...
+	uint8_t *tag = hw_ste->tag;
+
+	if (DR_STE_IS_OUTER_MPLS_OVER_GRE_SET(misc_2_mask)) {
+		DR_STE_SET_TAG(flex_parser_0, tag, parser_3_label,
+			       misc_2_mask, outer_first_mpls_over_gre_label);
+
+		DR_STE_SET_TAG(flex_parser_0, tag, parser_3_exp,
+			       misc_2_mask, outer_first_mpls_over_gre_exp);
+
+
```
Comment (jgunthorpe) [inline providers/mlx5/dr_ste.c]: redundant ()

**Comment by DistantNight** [comment]
> which depends on some kernel code to be accepted
I met some trouble when attach the mellanox nic to ovs-dpdk bridge using kernel 4.19
I want to know the effected kernel commit

**Pull request merged**